File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>
using namespace std;

bool visited[11][11][11][11] = { false };  // from (y1,x1) to (y2,x2)

int solution(string dirs) 
{
    int answer = 0;
    int curY = 5;
    int curX = 5;

    for (char dir : dirs)
    {
        int nextY = curY;
        int nextX = curX;

        if (dir == 'U') nextY--;
        if (dir == 'D') nextY++;
        if (dir == 'L') nextX--;
        if (dir == 'R') nextX++;

        if (nextY < 0 || nextY > 10 || nextX < 0 || nextX > 10)
            continue;

        if (!visited[curY][curX][nextY][nextX]) {
            visited[curY][curX][nextY][nextX] = true;
            visited[nextY][nextX][curY][curX] = true;
            answer++;
        }

        curY = nextY;
        curX = nextX;
    }

    return answer;
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int solution(string dirs) 
    {
        HashSet<(int a, int b, int c, int d)> visited = new HashSet<(int, int, int, int)>();

        int curY = 0,curX = 0;
        
        foreach(char dir in dirs)
        {
            int tempY = curY;
            int tempX = curX;
            
            if(dir == 'U') tempY++;
            else if(dir =='D') tempY--;
            else if(dir == 'R')tempX++;
            else if(dir == 'L')tempX--;
            
            if(tempY > 5 || tempY < -5 || tempX > 5 || tempX < -5) continue;
            
            visited.Add((tempY,tempX,curY,curX));
            visited.Add((curY,curX,tempY,tempX));
            
            curY = tempY;
            curX = tempX;
        }
        
        int answer = visited.Count / 2;
        return answer;
    }
}